File: src/cloudtwin/providers/azure/servicebus/service.py
"""Azure Service Bus – pure business logic."""

from __future__ import annotations

import logging
import uuid
from datetime import datetime, timezone
from typing import Optional

from cloudtwin.core.errors import NotFoundError
from cloudtwin.core.telemetry import TelemetryEngine
from cloudtwin.persistence.models.azure import AsbMessage, AsbQueue, AsbSubscription, AsbTopic
from cloudtwin.persistence.repositories.azure import (
    AsbMessageRepository,
    AsbQueueRepository,
    AsbSubscriptionRepository,
    AsbTopicRepository,
)

log = logging.getLogger("cloudtwin.azure.servicebus")


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


class ServiceBusService:
    def __init__(
        self,
        namespace: str,
        queue_repo: AsbQueueRepository,
        topic_repo: AsbTopicRepository,
        subscription_repo: AsbSubscriptionRepository,
        message_repo: AsbMessageRepository,
        telemetry: TelemetryEngine,
    ):
        self._ns = namespace
        self._queues = queue_repo
        self._topics = topic_repo
        self._subscriptions = subscription_repo
        self._messages = message_repo
        self._telemetry = telemetry

    # ------------------------------------------------------------------
    # Queues
    # ------------------------------------------------------------------

    async def create_queue(self, name: str) -> AsbQueue:
        existing = await self._queues.get(self._ns, name)
        if existing:
            return existing
        queue = AsbQueue(namespace=self._ns, name=name, created_at=_now())
        result = await self._queues.save(queue)
        await self._telemetry.emit("azure", "servicebus", "create_queue", {"queue": name})
        return result

    async def get_queue(self, name: str) -> AsbQueue:
        queue = await self._queues.get(self._ns, name)
        if not queue:
            raise NotFoundError(f"Queue not found: {name}")
        return queue

    async def list_queues(self) -> list[AsbQueue]:
        return await self._queues.list_by_namespace(self._ns)

    async def delete_queue(self, name: str) -> None:
        await self.get_queue(name)
        await self._queues.delete(self._ns, name)
        await self._telemetry.emit("azure", "servicebus", "delete_queue", {"queue": name})

    # ------------------------------------------------------------------
    # Topics
    # ------------------------------------------------------------------

    async def create_topic(self, name: str) -> AsbTopic:
        existing = await self._topics.get(self._ns, name)
        if existing:
            return existing
        topic = AsbTopic(namespace=self._ns, name=name, created_at=_now())
        result = await self._topics.save(topic)
        await self._telemetry.emit("azure", "servicebus", "create_topic", {"topic": name})
        return result

    async def get_topic(self, name: str) -> AsbTopic:
        topic = await self._topics.get(self._ns, name)
        if not topic:
            raise NotFoundError(f"Topic not found: {name}")
        return topic

    async def list_topics(self) -> list[AsbTopic]:
        return await self._topics.list_by_namespace(self._ns)

    async def delete_topic(self, name: str) -> None:
        await self.get_topic(name)
        await self._topics.delete(self._ns, name)
        await self._telemetry.emit("azure", "servicebus", "delete_topic", {"topic": name})

    # ------------------------------------------------------------------
    # Subscriptions
    # ------------------------------------------------------------------

    async def create_subscription(self, topic_name: str, sub_name: str) -> AsbSubscription:
        topic = await self.get_topic(topic_name)
        existing = await self._subscriptions.get(topic.id, sub_name)
        if existing:
            return existing
        sub = AsbSubscription(topic_id=topic.id, name=sub_name, created_at=_now())
        result = await self._subscriptions.save(sub)
        await self._telemetry.emit("azure", "servicebus", "create_subscription", {
            "topic": topic_name, "subscription": sub_name
        })
        return result

    async def get_subscription(self, topic_name: str, sub_name: str) -> AsbSubscription:
        topic = await self.get_topic(topic_name)
        sub = await self._subscriptions.get(topic.id, sub_name)
        if not sub:
            raise NotFoundError(f"Subscription not found: {topic_name}/{sub_name}")
        return sub

    async def list_subscriptions(self, topic_name: str) -> list[AsbSubscription]:
        topic = await self.get_topic(topic_name)
        return await self._subscriptions.list_by_topic(topic.id)

    async def delete_subscription(self, topic_name: str, sub_name: str) -> None:
        sub = await self.get_subscription(topic_name, sub_name)
        await self._subscriptions.delete(sub.topic_id, sub_name)

    # ------------------------------------------------------------------
    # Messages – queues
    # ------------------------------------------------------------------

    async def send_to_queue(self, queue_name: str, body: str, content_type: str = "text/plain") -> AsbMessage:
        queue = await self.get_queue(queue_name)
        msg = AsbMessage(
            message_id=str(uuid.uuid4()),
            entity_id=queue.id,
            entity_type="queue",
            body=body,
            content_type=content_type,
            lock_token=str(uuid.uuid4()),
            state="active",
            delivery_count=0,
            created_at=_now(),
        )
        result = await self._messages.save(msg)
        await self._telemetry.emit("azure", "servicebus", "send_queue_message", {"queue": queue_name})
        return result

    async def receive_from_queue(self, queue_name: str, limit: int = 1) -> list[AsbMessage]:
        queue = await self.get_queue(queue_name)
        messages = await self._messages.get_active(queue.id, "queue", limit=limit)
        for msg in messages:
            await self._messages.update_state(msg.lock_token, "locked")
            msg.state = "locked"
        return messages

    # ------------------------------------------------------------------
    # Messages – topics / subscriptions
    # ------------------------------------------------------------------

    async def send_to_topic(self, topic_name: str, body: str, content_type: str = "text/plain") -> list[AsbMessage]:
        topic = await self.get_topic(topic_name)
        subs = await self._subscriptions.list_by_topic(topic.id)
        results = []
        for sub in subs:
            msg = AsbMessage(
                message_id=str(uuid.uuid4()),
                entity_id=sub.id,
                entity_type="subscription",
                body=body,
                content_type=content_type,
                lock_token=str(uuid.uuid4()),
                state="active",
                delivery_count=0,
                created_at=_now(),
            )
            results.append(await self._messages.save(msg))
        await self._telemetry.emit("azure", "servicebus", "send_topic_message", {
            "topic": topic_name, "fan_out": len(results)
        })
        return results

    async def receive_from_subscription(
        self, topic_name: str, sub_name: str, limit: int = 1
    ) -> list[AsbMessage]:
        sub = await self.get_subscription(topic_name, sub_name)
        messages = await self._messages.get_active(sub.id, "subscription", limit=limit)
        for msg in messages:
            await self._messages.update_state(msg.lock_token, "locked")
            msg.state = "locked"
        return messages

    # ------------------------------------------------------------------
    # Message settlement
    # ------------------------------------------------------------------

    async def complete_message(self, lock_token: str) -> None:
        msg = await self._messages.get_by_lock_token(lock_token)
        if not msg:
            raise NotFoundError(f"Message not found: {lock_token}")
        await self._messages.delete(lock_token)

    async def abandon_message(self, lock_token: str) -> None:
        msg = await self._messages.get_by_lock_token(lock_token)
        if not msg:
            raise NotFoundError(f"Message not found: {lock_token}")
        await self._messages.update_state(lock_token, "active")

    async def deadletter_message(self, lock_token: str) -> None:
        msg = await self._messages.get_by_lock_token(lock_token)
        if not msg:
            raise NotFoundError(f"Message not found: {lock_token}")
        await self._messages.update_state(lock_token, "deadletter")

File: rust/cloudtwin-lite/src/azure/servicebus/service.rs
//! Azure Service Bus business logic.
//!
//! This is the provider-specific service layer for queues, topics,
//! subscriptions, and queue messages.

use anyhow::{bail, Result};
use chrono::Utc;
use rusqlite::OptionalExtension;
use uuid::Uuid;

use crate::db::Database;
use super::models::{AsbMessage, AsbQueue, AsbSubscription, AsbTopic};

/// Implements a simplified Service Bus over SQLite.
pub struct ServiceBusService {
    db:        Database,
    namespace: String,
}

impl ServiceBusService {
    pub fn new(db: Database, namespace: String) -> Self {
        ServiceBusService { db, namespace }
    }

    // ── Queues ────────────────────────────────────────────────────────────────

    pub async fn create_queue(&self, name: &str) -> Result<AsbQueue> {
        let n   = name.to_string();
        let ns  = self.namespace.clone();
        let now = Utc::now().to_rfc3339();
        let n2  = n.clone();
        let ns2 = ns.clone();
        let n3  = now.clone();
        self.db.conn.call(move |conn| {
            conn.execute(
                "INSERT INTO asb_queues (name, namespace, created_at)
                 VALUES (?1, ?2, ?3)
                 ON CONFLICT(namespace, name) DO NOTHING",
                rusqlite::params![n, ns, now],
            )?;
            Ok(())
        }).await.map_err(|e| anyhow::anyhow!("{e}"))?;
        Ok(AsbQueue { name: n2, namespace: ns2, created_at: n3 })
    }

    pub async fn list_queues(&self) -> Result<Vec<AsbQueue>> {
        let ns = self.namespace.clone();
        self.db.conn.call(move |conn| {
            let mut stmt = conn.prepare(
                "SELECT name, namespace, created_at FROM asb_queues
                 WHERE namespace=?1 ORDER BY name"
            )?;
            let rows = stmt.query_map(rusqlite::params![ns], |r| Ok(AsbQueue {
                name:       r.get(0)?,
                namespace:  r.get(1)?,
                created_at: r.get(2)?,
            }))?.collect::<rusqlite::Result<Vec<_>>>()?;
            Ok(rows)
        }).await.map_err(|e| anyhow::anyhow!("{e}"))
    }

    pub async fn delete_queue(&self, name: &str) -> Result<()> {
        let n  = name.to_string();
        let ns = self.namespace.clone();
        self.db.conn.call(move |conn| {
            conn.execute(
                "DELETE FROM asb_queues WHERE namespace=?1 AND name=?2",
                rusqlite::params![ns, n],
            )?;
            Ok(())
        }).await.map_err(|e| anyhow::anyhow!("{e}"))
    }

    // ── Topics ────────────────────────────────────────────────────────────────

    pub async fn create_topic(&self, name: &str) -> Result<AsbTopic> {
        let n   = name.to_string();
        let ns  = self.namespace.clone();
        let now = Utc::now().to_rfc3339();
        let n2  = n.clone();
        let ns2 = ns.clone();
        let n3  = now.clone();
        self.db.conn.call(move |conn| {
            conn.execute(
                "INSERT INTO asb_topics (name, namespace, created_at)
                 VALUES (?1, ?2, ?3)
                 ON CONFLICT(namespace, name) DO NOTHING",
                rusqlite::params![n, ns, now],
            )?;
            Ok(())
        }).await.map_err(|e| anyhow::anyhow!("{e}"))?;
        Ok(AsbTopic { name: n2, namespace: ns2, created_at: n3 })
    }

    pub async fn list_topics(&self) -> Result<Vec<AsbTopic>> {
        let ns = self.namespace.clone();
        self.db.conn.call(move |conn| {
            let mut stmt = conn.prepare(
                "SELECT name, namespace, created_at FROM asb_topics
                 WHERE namespace=?1 ORDER BY name"
            )?;
            let rows = stmt.query_map(rusqlite::params![ns], |r| Ok(AsbTopic {
                name:       r.get(0)?,
                namespace:  r.get(1)?,
                created_at: r.get(2)?,
            }))?.collect::<rusqlite::Result<Vec<_>>>()?;
            Ok(rows)
        }).await.map_err(|e| anyhow::anyhow!("{e}"))
    }

    pub async fn delete_topic(&self, name: &str) -> Result<()> {
        let n  = name.to_string();
        let ns = self.namespace.clone();
        self.db.conn.call(move |conn| {
            conn.execute(
                "DELETE FROM asb_topics WHERE namespace=?1 AND name=?2",
                rusqlite::params![ns, n],
            )?;
            Ok(())
        }).await.map_err(|e| anyhow::anyhow!("{e}"))
    }

    // ── Subscriptions ─────────────────────────────────────────────────────────

    pub async fn create_subscription(&self, topic: &str, name: &str) -> Result<AsbSubscription> {
        let n   = name.to_string();
        let t   = topic.to_string();
        let ns  = self.namespace.clone();
        let now = Utc::now().to_rfc3339();
        let n2  = n.clone();
        let t2  = t.clone();
        let ns2 = ns.clone();
        let n3  = now.clone();
        self.db.conn.call(move |conn| {
            conn.execute(
                "INSERT INTO asb_subscriptions (name, topic, namespace, created_at)
                 VALUES (?1, ?2, ?3, ?4)
                 ON CONFLICT(namespace, topic, name) DO NOTHING",
                rusqlite::params![n, t, ns, now],
            )?;
            Ok(())
        }).await.map_err(|e| anyhow::anyhow!("{e}"))?;
        Ok(AsbSubscription { name: n2, topic: t2, namespace: ns2, created_at: n3 })
    }

    pub async fn list_subscriptions(&self, topic: &str) -> Result<Vec<AsbSubscription>> {
        let t  = topic.to_string();
        let ns = self.namespace.clone();
        self.db.conn.call(move |conn| {
            let mut stmt = conn.prepare(
                "SELECT name, topic, namespace, created_at FROM asb_subscriptions
                 WHERE namespace=?1 AND topic=?2 ORDER BY name"
            )?;
            let rows = stmt.query_map(rusqlite::params![ns, t], |r| Ok(AsbSubscription {
                name:       r.get(0)?,
                topic:      r.get(1)?,
                namespace:  r.get(2)?,
                created_at: r.get(3)?,
            }))?.collect::<rusqlite::Result<Vec<_>>>()?;
            Ok(rows)
        }).await.map_err(|e| anyhow::anyhow!("{e}"))
    }

    // ── Messages ──────────────────────────────────────────────────────────────

    pub async fn send_message(&self, queue_name: &str, body: &str) -> Result<String> {
        let mid  = Uuid::new_v4().to_string();
        let lt   = Uuid::new_v4().to_string();
        let qn   = queue_name.to_string();
        let ns   = self.namespace.clone();
        let b    = body.to_string();
        let mid2 = mid.clone();
        let lt2  = lt.clone();
        let now  = Utc::now().to_rfc3339();
        self.db.conn.call(move |conn| {
            conn.execute(
                "INSERT INTO asb_messages
                 (message_id, namespace, queue_name, body, lock_token, visible, created_at)
                 VALUES (?1, ?2, ?3, ?4, ?5, 1, ?6)",
                rusqlite::params![mid2, ns, qn, b, lt2, now],
            )?;
            Ok(())
        }).await.map_err(|e| anyhow::anyhow!("{e}"))?;
        Ok(mid)
    }

    pub async fn receive_messages(&self, queue_name: &str, max: i64) -> Result<Vec<AsbMessage>> {
        let qn  = queue_name.to_string();
        let ns  = self.namespace.clone();
        let lim = max.min(10).max(1);
        let msgs: Vec<AsbMessage> = self.db.conn.call(move |conn| {
            let mut stmt = conn.prepare(
                "SELECT message_id, namespace, queue_name, body, lock_token, visible, created_at
                 FROM asb_messages
                 WHERE namespace=?1 AND queue_name=?2 AND visible=1
                 ORDER BY id LIMIT ?3"
            )?;
            let rows = stmt.query_map(rusqlite::params![ns, qn, lim], |r| Ok(AsbMessage {
                message_id:  r.get(0)?,
                namespace:   r.get(1)?,
                queue_name:  r.get(2)?,
                body:        r.get(3)?,
                lock_token:  r.get(4)?,
                visible:     r.get::<_, i64>(5)? != 0,
                created_at:  r.get(6)?,
            }))?.collect::<rusqlite::Result<Vec<_>>>()?;
            Ok(rows)
        }).await.map_err(|e| anyhow::anyhow!("{e}"))?;

        for msg in &msgs {
            let lt = msg.lock_token.clone();
            self.db.conn.call(move |conn| {
                conn.execute(
                    "UPDATE asb_messages SET visible=0 WHERE lock_token=?1",
                    rusqlite::params![lt],
                )?;
                Ok(())
            }).await.map_err(|e| anyhow::anyhow!("{e}"))?;
        }
        Ok(msgs)
    }

    pub async fn complete_message(&self, lock_token: &str) -> Result<()> {
        let lt = lock_token.to_string();
        self.db.conn.call(move |conn| {
            conn.execute(
                "DELETE FROM asb_messages WHERE lock_token=?1",
                rusqlite::params![lt],
            )?;
            Ok(())
        }).await.map_err(|e| anyhow::anyhow!("{e}"))
    }

    pub async fn abandon_message(&self, lock_token: &str) -> Result<()> {
        let lt = lock_token.to_string();
        self.db.conn.call(move |conn| {
            conn.execute(
                "UPDATE asb_messages SET visible=1 WHERE lock_token=?1",
                rusqlite::params![lt],
            )?;
            Ok(())
        }).await.map_err(|e| anyhow::anyhow!("{e}"))
    }

    pub async fn send_topic_message(&self, topic: &str, body: &str) -> Result<String> {
        // Fan out to all subscriptions treated as virtual queues
        self.send_message(&format!("topic:{topic}"), body).await
    }

    pub async fn receive_subscription_messages(
        &self,
        topic: &str,
        subscription: &str,
        max: i64,
    ) -> Result<Vec<AsbMessage>> {
        let virtual_queue = format!("topic:{topic}:sub:{subscription}");
        self.receive_messages(&virtual_queue, max).await
    }
}
